File: src/interface/Python/paramonte/_AutoCorr.py
####################################################################################################################################
####################################################################################################################################
####
####   MIT License
####
####   ParaMonte: plain powerful parallel Monte Carlo library.
####
####   Copyright (C) 2012-present, The Computational Data Science Lab
####
####   This file is part of the ParaMonte library.
####
####   Permission is hereby granted, free of charge, to any person obtaining a 
####   copy of this software and associated documentation files (the "Software"), 
####   to deal in the Software without restriction, including without limitation 
####   the rights to use, copy, modify, merge, publish, distribute, sublicense, 
####   and/or sell copies of the Software, and to permit persons to whom the 
####   Software is furnished to do so, subject to the following conditions:
####
####   The above copyright notice and this permission notice shall be 
####   included in all copies or substantial portions of the Software.
####
####   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
####   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
####   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
####   IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, 
####   DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
####   OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE 
####   OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
####
####   ACKNOWLEDGMENT
####
####   ParaMonte is an honor-ware and its currency is acknowledgment and citations.
####   As per the ParaMonte library license agreement terms, if you use any parts of 
####   this library for any purposes, kindly acknowledge the use of ParaMonte in your 
####   work (education/research/industry/development/...) by citing the ParaMonte 
####   library as described on this page:
####
####       https://github.com/cdslaborg/paramonte/blob/master/ACKNOWLEDGMENT.md
####
####################################################################################################################################
####################################################################################################################################

import numpy as np
import typing as tp
import pandas as pd
import weakref as wref

import _message as err
import _dfutils as dfutils
import _pmutils as pmutils
from paramonte.vis.LineScatterPlot import LineScatterPlot

Struct = pmutils.Struct
newline = pmutils.newline

####################################################################################################################################
#### AutoCorr class
####################################################################################################################################

class AutoCorr:
    """

    This is the base class for generating object of type ``AutoCorr``.  
    Upon construction, it will provide methods to compute and plot the 
    autocorrelations of the selected columns of the input dataFrame.

        **Parameters**

            dataFrame

                A Pandas dataframe based upon the selected columns 
                of which the autocorrelations will be computed.

            columns (optional)

                A list that determines the columns of the input dataFrame 
                to be used in the computation of the autocorrelations. 
                It can have three forms:

                    1.  A list of column indices from the input dataFrame.
                    2.  A list of column names from dataFrame.columns.
                    3.  A ``range(start,stop,step)`` of column indices.

                Examples:

                    1.  ``columns = [0,1,4,3]``
                    2.  ``columns = ["SampleLogFunc","SampleVariable1"]``
                    3.  ``columns = range(17,7,-2)``

                The default behavior includes all columns of the dataFrame.

        **Attributes**

            All of the parameters described above, except dataFrame.

                A weak reference to the dataFrame will be implicitly 
                stored in the object. 

            df

                A pandas dataframe containing the computed autocorrelations.

            rows

                A list of the dataFrame indices that determines the rows of 
                the input dataFrame to be used in the computation of the 
                autocorrelations. It can be either:

                    1.  A ``range(start,stop,step)``, or, 
                    2.  A list of row indices from ``dataFrame.index``.

                Examples:

                    1.  ``rows = range(17,7,-2)``
                    2.  ``rows = [i for i in range(7,17)]``

                The default behavior includes all rows of the dataFrame.

            plot

                A structure containing the following plotting tools:

                    heatmap

                        A callable object of class HeatMap which will 
                        enable plotting of the correlation matrix.

        **Returns**

            self

                an object of type class CorMat.

    ---------------------------------------------------------------------------
    """

    ################################################################################################################################
    #### __init__
    ################################################################################################################################

    def __init__( self
                , dataFrame     # : tp.Optional[ pd.DataFrame ]
                , columns       = None # : tp.Optional[ tp.Union[ range , tp.List[int] , tp.List[str] ] ] = None
                , methodName    = "ParaMonte" # : tp.Optional[ str ] = "ParaMonte"
                , reportEnabled = True # : tp.Optional[ bool ] = True
                ):

        self.df = None
        self.rows = None
        self.columns = columns

        self._dfref = None if dataFrame is None else wref.ref(dataFrame)

        self._methodName = methodName
        self._reportEnabled = reportEnabled
        self._progress = pmutils.Progress   ( msg = None
                                            , methodName = methodName
                                            , reportEnabled = reportEnabled
                                            , end = "\n"
                                            )

    ################################################################################################################################
    #### __call__
    ################################################################################################################################

    def __call__( self
                , reself : tp.Optional[ bool ] = False
                , **kwargs
                ):
        """

        Call the ``get()`` method of the current instance of the class.

            **Parameters**

                Any arguments that can be passed to the ``get()`` 
                method of the object.

            **Returns**

                Any return value from the ``get()`` method of the object.

        """

        return self.get(reself, **kwargs)

    ################################################################################################################################
    #### get
    ################################################################################################################################

    def get ( self
            , reself : tp.Optional[ bool ] = False
            , **kwargs
            ):
        """

        Compute the autocorrelations of the selected columns 
        of the input dataframe to the object's constructor.

            **Parameters**

                reself

                    A logical variable. When ``True``, an instance of 
                    the object will be returned to the calling routine 
                    upon exit. The default value is ``False``.

            **Returns**

                The object if ``reself = True`` otherwise, ``None``.

                **NOTE**

                This method causes side-effects by manipulating
                the existing attributes of the object.

        """

        for key in kwargs.keys():
            if hasattr(self,key):
                setattr(self, key, kwargs[key])
            elif key=="dataFrame":
                setattr( self, "_dfref", wref.ref(kwargs[key]) )
            else:
                raise Exception ( "Unrecognized input '"+key+"' class attribute detected." + newline
                                + self._getDocString()
                                )

        ############################################################################################################################
        #### check columns presence
        ############################################################################################################################

        colnames, colindex = dfutils.getColNamesIndex(self._dfref().columns,self.columns)

        ############################################################################################################################
        #### check rows presence. This must be checked here, because it depends on the integrity of the in input dataFrame.
        ############################################################################################################################

        if self.rows is None: self.rows = range(len(self._dfref().index))
        #rownames = self._dfref().index[self.rows]

        ############################################################################################################################
        #### compute the autocorrelations
        ############################################################################################################################

        self._nvar = len(colnames)
        self._nlag = len(self.rows)
        acf = np.zeros((self._nvar,self._nlag))

        try:

            from scipy.signal import correlate
            for cnt, ivar in enumerate(colindex):
                xdata = self._dfref().iloc[self.rows,ivar].values.flatten() - np.mean(self._dfref().iloc[self.rows,ivar].values.flatten())
                acf[cnt] = correlate( xdata
                                    , xdata
                                    , mode = "full"
                                    )[self._nlag-1:2*self._nlag]
                acf[cnt] = acf[cnt] / acf[cnt,0]

        except:

            if self._reportEnabled:
                err.warn( msg   = "Failed to compute the Autocorrelation function of the input dataFrame." + newline
                                + "This could likely be due to an error in importing the scipy Python library." + newline
                                + "Please make sure you have the scipy library properly installed on your system." + newline
                                + "You can do so by typing the following command on your Anaconda3 or Bash command prompt:" + newline
                                + newline
                                + "    pip install --user --upgrade scipy"
                        , methodName = self._methodName
                        , marginTop = 1
                        , marginBot = 1
                        )

        self.df = pd.DataFrame(np.transpose(acf))

        ############################################################################################################################
        #### specify columns/index names
        ############################################################################################################################

        colnames = [ "ACF_"+colnames[i] for i in range(len(colnames)) ]
        self.df.columns = colnames

        ############################################################################################################################
        #### add SampleLogFunc to df for plot coloring, if exists
        ############################################################################################################################

        ccolumns = ()
        if "SampleLogFunc" in self._dfref().columns:
            ccolumns = "SampleLogFunc"
            self.df.insert  ( loc = 0
                            , column = ccolumns
                            , value = self._dfref()[[ccolumns]].values.flatten()
                            , allow_duplicates = True
                            )

        ############################################################################################################################
        #### add lags to df
        ############################################################################################################################

        self.df.insert  ( loc = 0
                        , column = "Lag"
                        , value = [ i for i in self.df.index ]
                        )

        ############################################################################################################################
        #### graphics
        ############################################################################################################################

        self._plotTypeList =    [ "line"
                                , "scatter"
                                , "lineScatter"
                                ]

        self._progress.note( msg = "adding the autocrrelation graphics tools... ", end = newline, pre = True )
        self.plot = Struct()
        self._resetPlot(resetType="hard")

        self.plot.reset = self._resetPlot

        ############################################################################################################################

        if reself: return self

    ################################################################################################################################
    #### _resetPlot
    ################################################################################################################################

    def _resetPlot  ( self
                    , resetType = "soft"
                    , plotNames = "all"
                    ):
        """

        Reset the properties of the plot to the original default settings.
        Use this method when you change many attributes of the plot and
        you want to clean up and go back to the default settings.

            **Parameters**

                resetType (optional)

                    An optional string with possible value of ``"hard"``.
                    If provided, the plot object will be regenerated from scratch.
                    This includes reading the original data frame again and resetting
                    everything. If not provided, then only the plot settings will be
                    reset without reseting the dataFrame.

                plotNames (optional)

                    An optional string value or list of string values representing 
                    the names of plots to reset. If no value is provided, 
                    then all plots will be reset.

            **Returns**

                None

            **Example**

                .. code-block:: python

                    reset("hard")                    # regenerate all plots from scratch
                    reset("hard","line")             # regenerate line plot from scratch
                    reset("hard",["line","scatter"]) # regenerate line and scatter plots

        """

        if isinstance(plotNames, str):
            plotTypeLower = plotNames.lower()
            if plotTypeLower=="all":
                requestedPlotTypeList = self._plotTypeList
            elif plotNames in self._plotTypeList:
                requestedPlotTypeList = [plotNames]
            else:
                self._reportWrongPlotName(plotNames)
        elif isinstance(plotNames, list):
            for plotName in plotNames:
                if plotName not in self._plotTypeList: self._reportWrongPlotName(plotName)
        else:
            self._reportWrongPlotName("a none-string none-list object.")

        resetTypeIsHard = None
        if isinstance(resetType, str):
            resetTypeIsHard = resetType.lower()=="hard"
        else:
            err.abort   ( msg   = "The input argument resetType must be a string representing" + newline
                                + "the type of the reset to be performed on the plots." + newline
                                + "A list of possible plots includes: \"hard\", \"soft\"" + newline
                                + "Here is the help for the ``reset()`` method: " + newline
                                + newline
                                + self._resetPlot.__doc__
                        , marginTop = 1
                        , marginBot = 1
                        , methodName = self._methodName
                        )

        ############################################################################################################################
        #### reset plots
        ############################################################################################################################

        for requestedPlotType in requestedPlotTypeList:

            plotObject = None
            requestedPlotTypeLower = requestedPlotType.lower()

            isLine      = "line"        in requestedPlotTypeLower
            isScatter   = "scatter"     in requestedPlotTypeLower

            isLineScatterPlot = isLine or isScatter

            if not resetTypeIsHard:
                plotComponent = getattr(self, "plot")
                plotObject = getattr(plotComponent, requestedPlotType)
                plotObject._reset()

            ########################################################################################################################
            #### reset line / scatter
            ########################################################################################################################

            if isLineScatterPlot:

                if resetTypeIsHard:
                    plotObject = LineScatterPlot( plotType = requestedPlotType
                                                , dataFrame = self.df
                                                , methodName = self._methodName
                                                , reportEnabled = self._reportEnabled
                                                , resetPlot = self._resetPlot
                                                )

                plotObject.xcolumns = "Lag"
                plotObject.ycolumns = self.df.columns[2:]

                plotObject._xlabel = "AutoCorrelation Lag"
                plotObject._ylabel = "AutoCorrelation Function (ACF) Value"

                plotObject.ccolumns = None

                plotObject.colorbar.kws.extend = "neither"
                plotObject.colorbar.kws.orientation = "vertical"
                plotObject.colorbar.kws.spacing = "uniform"

                plotObject._xlimit = [1, None]
                plotObject._ylimit = [None, 1]
                plotObject._xscale = "log"

                plotObject.legend.enabled = True

                if isLine:
                    if isScatter:
                        plotObject.lineCollection.enabled = False
                        plotObject.plot.enabled = True
                        plotObject.plot.kws.alpha = 0.2
                        plotObject.plot.kws.color = "grey"
                        plotObject.plot.kws.linewidth = 0.75
                    else:
                        plotObject.plot.enabled = True
                        plotObject.plot.kws.linewidth = 1
                        plotObject.lineCollection.enabled = False

            ########################################################################################################################

            if plotObject is not None: setattr(self.plot, requestedPlotType, plotObject)

    ################################################################################################################################
    #### _reportWrongPlotName
    ################################################################################################################################

    def _reportWrongPlotName( self
                            , plotNames
                            ):

        err.abort   ( msg   = "The input argument plotNames must be a string representing" + newline
                            + "the name of a plot belonging to the AutoCorr class or," + newline
                            + "a list of such plot names. You have entered: " + plotNames + newline
                            + "Possible plots are: " + newline
                            + newline
                            + newline.join(self._plotTypeList) + newline
                            + newline
                            + "Here is the help for the ``reset()`` method: " + newline
                            + newline
                            + self._resetPlot.__doc__
                    , marginTop = 1
                    , marginBot = 1
                    , methodName = self._methodName
                    )

    ################################################################################################################################

File: src/interface/MATLAB/paramonte/stats/AutoCorr.m
classdef AutoCorr < dynamicprops
%   This is the class for generating object of type CorMat which, 
%   upon construction, will provide methods to compute and plot the 
%   autocorrelations of the selected columns of the input dataFrame.
%
%   Parameters
%   ----------
%       dataFrame
%           a Pandas dataframe based upon the selected comlumns of which 
%           the autocorrelations will be computed.
%       columns
%           optional argument that determines the columns of the input dataFrame to be 
%           used in the computation of the autocorrelations. It can have three forms:
%               1. a list of column indices from the input dataFrame.
%               2. a list of column names from dataFrame.columns.
%               3. a range(start,stop,step), representing the column indices in dataFrame.
%           Examples:
%               1. columns = [0,1,4,3]
%               2. columns = ["SampleLogFunc","SampleVariable1"]
%               3. columns = range(17,7,-2)
%           If not provided, the default behavior includes all columns of the dataFrame.
%       rows
%           optional argument that determines the rows of the input dataFrame to be 
%           used in the computation of the autocorrelations. It can be either:
%               1. a range(start,stop,step), or, 
%               2. a list of row indices from dataFrame.index.
%           Examples:
%               1. rows = range(17,7,-2)
%               2. rows = [i for i in range(7,17)]
%           If not provided, the default behavior includes all rows of the dataFrame.
%
%   Attributes
%   ----------
%       All of the parameters described above, except dataFrame.
%           a reference to the dataFrame will be implicitly stored in the object.
%       df
%           a pandas dataframe containing the computed autocorrelations.
%       plot
%           a structure containing the following plotting tools:
%               heatmap
%                   a callable object of class HeatMap which will enable 
%                   plotting of the correlation matrix.
%
%   Returns
%   -------
%       corMat
%           an object of type class CorMat.
%
%   ----------------------------------------------------------------------


    properties(Access = public)
        columns
        bounds
        rows
        df
    end

    properties(Hidden)
        plotTypeList = ["line","scatter","lineScatter"];
        cormatPrecision
        matrixType
        isCorMat
        offset
        dfref
        title
        Err
    end

    %*******************************************************************************************************************************
    %*******************************************************************************************************************************

    methods(Access=public)

        %***************************************************************************************************************************
        %***************************************************************************************************************************

        function self = AutoCorr    ( dataFrame ...
                                    , columns ...
                                    , rows ...
                                    , Err ...
                                    )
            self.Err = Err;
            self.dfref = dataFrame;
            self.columns = columns;
            self.rows = []; if ~isempty(rows); self.rows = rows; end
            self.df = [];
            self.offset = 2;

            prop="plot"; if ~any(strcmp(properties(self),prop)); self.addprop(prop); end
            self.plot = struct();

            self.get();

        end

        %***************************************************************************************************************************
        %***************************************************************************************************************************

        function get(self,varargin)
        %   compute the correlation matrix of the selected columns 
        %   of the input dataframe to the object's constructor.
        %   
        %   Parameters
        %   ----------
        %       None
        %   
        %   Returns
        %   -------
        %       None. However, this method causes side-effects by manipulating 
        %       the existing attributes of the object.

            parseArgs(self,varargin{:});

            % check columns presence

            if getVecLen(self.columns)
                [colnames, colindex] = getColNamesIndex(self.dfref.Properties.VariableNames,self.columns);
            else
                colindex = 1:length(self.dfref.Properties.VariableNames);
                colnames = string(self.dfref.Properties.VariableNames);
            end

            % check rows presence

            if getVecLen(self.rows)
                rowindex = self.rows;
            else
                rowindex = 1:1:length(self.dfref{:,1});
            end

            % compute the autocorrelations

            nvar = length(colindex);
            nlag = length(rowindex)-1;
            acf = zeros(nlag,nvar);
            bounds_array = zeros(2,nvar);

            for i = 1:length(colindex)
                [acf(:,i),lags,bounds_array(:,i)] = autocorr( self.dfref{rowindex,colindex(i)}, nlag-1 );
            end
            self.df = array2table([lags,acf]);
            colnames = "ACF_" + colnames;
            self.df.Properties.VariableNames = ["Lag", colnames];
            self.df.Properties.RowNames = string(lags);
            self.bounds = array2table(bounds_array);
            self.bounds.Properties.VariableNames = colnames;
            self.bounds.Properties.RowNames = ["upperLimit","lowerLimit"];

            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            %%%% graphics
            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

            for plotType = self.plotTypeList
                self.resetPlot(plotType,"hard");
                %self.plot.(plotType).target.value = [1 ]
            end
            self.plot.reset = @self.resetPlot;
            self.plot.addBounds = @self.addBounds;

        end % get

        %***************************************************************************************************************************
        %***************************************************************************************************************************

    end

    %*******************************************************************************************************************************
    %*******************************************************************************************************************************

    methods (Access = public, Hidden)

        %***************************************************************************************************************************
        %***************************************************************************************************************************

        function addBounds(self,plotType)
            currentFig = [];
            msg1= "The method addBounds() optionally takes an input argument ""plotType"" which is a string containing " + newline ...
                + "the type of the figure to which the bounds will be added. Possible plotType options are: " + newline + newline ...
                + "    " + strjoin(self.plotTypeList,", ") + newline;
            if nargin==1
                currentFig = get(groot,'CurrentFigure');
                if isempty(currentFig)
                    error   ( newline ...
                            + "Could not find any active figure to which the Autocorrelation Significance Bounds could be added." + newline ...
                            + msg1 ...
                            + newline ...
                            );
                end
            elseif nargin==2
                currentFig = self.plot.(plotType).currentFig.gcf;
                set(0, "CurrentFigure", currentFig);
            end
            currentAxes = currentFig.CurrentAxes;
            xLimits = currentAxes.XLim;
            boundListLen = length(self.bounds.Properties.VariableNames);
            xval = mean(xLimits);
            boundList = zeros(2*boundListLen,2);
            for i = 1:boundListLen
                indx = 2*(i-1)+1;
                boundList(indx,:) = [xval self.bounds{1,i}];
                boundList(indx+1,:) = [xval self.bounds{2,i}];
            end
            self.plot.(plotType).target.vline_kws.enabled = false;
            self.plot.(plotType).target.scatter_kws.enabled = false;
            self.plot.(plotType).target.hline_kws.enabled = true;
            self.plot.(plotType).target.hline_kws.color = "black";%[150 150 150 150]/255;
            self.plot.(plotType).target.xlimits = [xLimits(1)+0.001*xLimits(1), xLimits(2)];
            self.plot.(plotType).target.value = boundList;
            self.plot.(plotType).target.plot();
        end

        %***************************************************************************************************************************
        %***************************************************************************************************************************

        function resetPlot(self,varargin)

            requestedPlotTypeList = [];

            if nargin==1
                requestedPlotTypeList = self.plotTypeList;
            else
                for requestedPlotTypeCell = varargin{1}
                    if isa(requestedPlotTypeCell,"cell")
                        requestedPlotType = string(requestedPlotTypeCell{1});
                    else
                        requestedPlotType = string(requestedPlotTypeCell);
                    end
                    plotTypeNotFound = true;
                    for plotTypeCell = self.plotTypeList
                        plotType = string(plotTypeCell{1});
                        if strcmp(plotType,requestedPlotType)
                            requestedPlotTypeList = [ requestedPlotTypeList , plotType ];
                            plotTypeNotFound = false;
                            break;
                        end
                    end
                    if plotTypeNotFound
                        error   ( newline ...
                                + "The input plot-type argument, " + varargin{1} + ", to the resetPlot method" + newline ...
                                + "did not match any plot type. Possible plot types include:" + newline ...
                                + "    " + strjoin(self.plotTypeList,", ") + newline ...
                                );
                    end
                end
            end

            resetTypeIsHard = false;
            if nargin==3 && strcmpi(varargin{2},"hard")
                resetTypeIsHard = true;
                msgPrefix = "creating the ";
                msgSuffix = " plot object from scrach...";
            else
                msgPrefix = "reseting the properties of the ";
                msgSuffix = " plot...";
            end

            self.Err.marginTop = 0;
            self.Err.marginBot = 0;

            for requestedPlotTypeCell = requestedPlotTypeList

                self.Err.msg = msgPrefix + requestedPlotType + msgSuffix;
                self.Err.note();

                requestedPlotType = string(requestedPlotTypeCell);
                requestedPlotTypeLower = lower(requestedPlotType);
                plotName = "";

                %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

                is3d = false;
                if contains(requestedPlotTypeLower,"3")
                    is3d = true;
                end

                % line

                if strcmp(requestedPlotTypeLower,"line") || strcmp(requestedPlotTypeLower,"line3")
                    plotName = "line"; if is3d; plotName = plotName + "3"; end
                    if resetTypeIsHard
                        self.plot.(plotName) = LineScatterPlot( self.df, plotName );
                    else
                        self.plot.(plotName).reset();
                    end
                    self.plot.(plotName).xcolumns = string(self.df.Properties.VariableNames(self.offset-1));
                    self.plot.(plotName).ycolumns = string(self.df.Properties.VariableNames(self.offset:end));
                    self.plot.(plotName).surface_kws.enabled = false;
                    self.plot.(plotName).ccolumns = "SampleLogFunc";
                    self.plot.(plotName).gca_kws.xscale = "linear";
                    self.plot.(plotName).plot_kws.linewidth = 1;
                    self.plot.(plotName).surface_kws.linewidth = 1;
                end

                % scatter / scatter3

                if strcmp(requestedPlotTypeLower,"scatter") || strcmp(requestedPlotTypeLower,"scatter3")
                    plotName = "scatter"; if is3d; plotName = plotName + "3"; end
                    if resetTypeIsHard
                        self.plot.(plotName) = LineScatterPlot( self.df, plotName );
                    else
                        self.plot.(plotName).reset();
                    end
                    self.plot.(plotName).ccolumns = string(self.df.Properties.VariableNames(self.offset));
                    self.plot.(plotName).xcolumns = string(self.df.Properties.VariableNames(self.offset-1));
                    self.plot.(plotName).ycolumns = string(self.df.Properties.VariableNames(self.offset:end));
                    self.plot.(plotName).gca_kws.xscale = "linear";
                    self.plot.(plotName).scatter_kws.size = 10;
                end

                % lineScatter / lineScatter3

                if strcmp(requestedPlotTypeLower,"linescatter") || strcmp(requestedPlotTypeLower,"linescatter3")
                    plotName = "lineScatter"; if is3d; plotName = plotName + "3"; end
                    if resetTypeIsHard
                        self.plot.(plotName) = LineScatterPlot( self.df, plotName );
                    else
                        self.plot.(plotName).reset();
                    end
                    self.plot.(plotName).surface_kws.enabled = false;
                    self.plot.(plotName).ccolumns = string(self.df.Properties.VariableNames(self.offset));
                    self.plot.(plotName).xcolumns = string(self.df.Properties.VariableNames(self.offset-1));
                    self.plot.(plotName).ycolumns = string(self.df.Properties.VariableNames(self.offset:end));
                    self.plot.(plotName).gca_kws.xscale = "linear";
                    if is3d
                        self.plot.(plotName).plot_kws.color = [200 200 200 75] / 255;
                    else
                        self.plot.(plotName).plot_kws.linewidth = 1;
                        self.plot.(plotName).plot_kws.color = uint8([200 200 200 100]);
                        self.plot.(plotName).scatter_kws.size = 20;
                    end
                end

                % 3d

                if is3d
                    self.plot.(plotName).xcolumns = string(self.df.Properties.VariableNames(self.offset));
                    self.plot.(plotName).ycolumns = string(self.df.Properties.VariableNames(self.offset+1));
                    self.plot.(plotName).zcolumns = string(self.df.Properties.VariableNames(self.offset));
                end

                %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

            end

        end

        %***************************************************************************************************************************
        %***************************************************************************************************************************

    end % methods

    %*******************************************************************************************************************************
    %*******************************************************************************************************************************

end
